<a href="https://colab.research.google.com/github/ady909/Capstone-Project-Bike-Sharing-Demand-Prediction-/blob/main/Capstone_Project_Bike_Sharing_Demand_Prediction_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Project Title : Seoul Bike Sharing Demand Prediction

### Project Type - Regression

### Contribution - Individual

## Problem Description

Currently Rental bikes are introduced in many urban cities for the enhancement of mobility comfort. It is important to make the rental bike available and accessible to the public at the right time as it lessens the waiting time. Eventually, providing the city with a stable supply of rental bikes becomes a major concern. The crucial part is the prediction of bike count required at each hour for the stable supply of rental bikes.

##Attribute Information:

Date : year-month-day

Rented Bike count - Count of bikes rented at each hour

Hour - Hour of he day

Temperature-Temperature in Celsius

Humidity - %

Windspeed - m/s

Visibility - 10m

Dew point temperature - Celsius

Solar radiation - MJ/m2

Rainfall - mm

Snowfall - cm

Seasons - Winter, Spring, Summer, Autumn

Holiday - Holiday/No holiday

Functional Day - NoFunc(Non Functional Hours), Fun(Functional hours)

In [1]:
from google.colab import files
upload = files.upload()

Saving SeoulBikeData.csv to SeoulBikeData (1).csv


In [2]:
#importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import datetime as dt
from sklearn.model_selection import train_test_split , GridSearchCV
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.linear_model import Ridge, Lasso, LinearRegression

import warnings
warnings.filterwarnings('ignore')

%matplotlib inline
sns.set_style("whitegrid",{'grid.linestyle': '--'})


In [3]:
#loading data
bike_df = pd.read_csv("SeoulBikeData.csv" , encoding = "unicode_escape")

## Getting to know about data

In [4]:
#first 20 rows look
bike_df.head(n = 20)

Date  Rented Bike Count  Hour  Temperature(°C)  Humidity(%)  \
0   01/12/2017                254     0             -5.2           37   
1   01/12/2017                204     1             -5.5           38   
2   01/12/2017                173     2             -6.0           39   
3   01/12/2017                107     3             -6.2           40   
4   01/12/2017                 78     4             -6.0           36   
5   01/12/2017                100     5             -6.4           37   
6   01/12/2017                181     6             -6.6           35   
7   01/12/2017                460     7             -7.4           38   
8   01/12/2017                930     8             -7.6           37   
9   01/12/2017                490     9             -6.5           27   
10  01/12/2017                339    10             -3.5           24   
11  01/12/2017                360    11             -0.5           21   
12  01/12/2017                449    12              1.7           23   
13  01/12/2017                451    13              2.4           25   
14  01/12/2017                447    14              3.0           26   
15  01/12/2017                463    15              2.1           36   
16  01/12/2017                484    16              1.2           54   
17  01/12/2017                555    17              0.8           58   
18  01/12/2017                862    18              0.6           66   
19  01/12/2017                600    19              0.0           77   

    Wind speed (m/s)  Visibility (10m)  Dew point temperature(°C)  \
0                2.2              2000                      -17.6   
1                0.8              2000                      -17.6   
2                1.0              2000                      -17.7   
3                0.9              2000                      -17.6   
4                2.3              2000                      -18.6   
5                1.5              2000                      -18.7   
6                1.3              2000                      -19.5   
7                0.9              2000                      -19.3   
8                1.1              2000                      -19.8   
9                0.5              1928                      -22.4   
10               1.2              1996                      -21.2   
11               1.3              1936                      -20.2   
12               1.4              2000                      -17.2   
13               1.6              2000                      -15.6   
14               2.0              2000                      -14.6   
15               3.2              2000                      -11.4   
16               4.2               793                       -7.0   
17               1.6              2000                       -6.5   
18               1.4              2000                       -5.0   
19               1.7              2000                       -3.5   

    Solar Radiation (MJ/m2)  Rainfall(mm)  Snowfall (cm) Seasons     Holiday  \
0                      0.00           0.0            0.0  Winter  No Holiday   
1                      0.00           0.0            0.0  Winter  No Holiday   
2                      0.00           0.0            0.0  Winter  No Holiday   
3                      0.00           0.0            0.0  Winter  No Holiday   
4                      0.00           0.0            0.0  Winter  No Holiday   
5                      0.00           0.0            0.0  Winter  No Holiday   
6                      0.00           0.0            0.0  Winter  No Holiday   
7                      0.00           0.0            0.0  Winter  No Holiday   
8                      0.01           0.0            0.0  Winter  No Holiday   
9                      0.23           0.0            0.0  Winter  No Holiday   
10                     0.65           0.0            0.0  Winter  No Holiday   
11                     0.94           0.0            0.0  Winter  

In [5]:
#last 20 rows look
bike_df.tail(n = 20)

Date  Rented Bike Count  Hour  Temperature(°C)  Humidity(%)  \
8740  30/11/2018                116     4             -0.5           71   
8741  30/11/2018                149     5             -0.7           66   
8742  30/11/2018                293     6             -0.8           68   
8743  30/11/2018                750     7             -1.2           70   
8744  30/11/2018               1527     8             -1.5           68   
8745  30/11/2018                809     9             -0.4           57   
8746  30/11/2018                554    10              1.9           51   
8747  30/11/2018                642    11              5.3           43   
8748  30/11/2018                720    12              6.6           35   
8749  30/11/2018                740    13              7.1           24   
8750  30/11/2018                761    14              7.8           20   
8751  30/11/2018                768    15              7.0           20   
8752  30/11/2018                837    16              7.2           23   
8753  30/11/2018               1047    17              6.0           29   
8754  30/11/2018               1384    18              4.7           34   
8755  30/11/2018               1003    19              4.2           34   
8756  30/11/2018                764    20              3.4           37   
8757  30/11/2018                694    21              2.6           39   
8758  30/11/2018                712    22              2.1           41   
8759  30/11/2018                584    23              1.9           43   

      Wind speed (m/s)  Visibility (10m)  Dew point temperature(°C)  \
8740               0.4              1345                       -5.1   
8741               0.5              1336                       -6.2   
8742               0.8              1322                       -5.9   
8743               0.8              1351                       -5.9   
8744               1.1              1286                       -6.6   
8745               0.6              1270                       -7.8   
8746               0.8              1029                       -7.1   
8747               1.8              1177                       -6.2   
8748               1.3              1409                       -7.8   
8749               2.8              1838                      -12.1   
8750               2.2              2000                      -13.8   
8751               3.3              1994                      -14.4   
8752               1.5              1945                      -12.6   
8753               2.1              1877                      -10.7   
8754               1.9              1661                       -9.8   
8755               2.6              1894                      -10.3   
8756               2.3              2000                       -9.9   
8757               0.3              1968                       -9.9   
8758               1.0              1859                       -9.8   
8759               1.3              1909                       -9.3   

      Solar Radiation (MJ/m2)  Rainfall(mm)  Snowfall (cm) Seasons  \
8740                     0.00           0.0            0.0  Autumn   
8741                     0.00           0.0            0.0  Autumn   
8742                     0.00           0.0            0.0  Autumn   
8743                     0.00           0.0            0.0  Autumn   
8744                     0.02           0.0            0.0  Autumn   
8745                     0.45           0.0            0.0  Autumn   
8746                     1.01           0.0            0.0  Autumn   
8747                     1.38           0.0            0.0  Autumn   
8748                     1.70           0.0            0.0  Autumn   
8749                     1.83           0.0            0.0  Autumn   
8750                     1.67           0.0            0.0  Autumn   
8751                     1.21           0.0            0.0  Autumn   
8752                     0.72     

Data in consist of info about the bike rented on hourly bases from 01/12/2017 to 30/11/2018. Weather conditions are also taken into consideration.

In [6]:
#chcking dimensions of dataset
bike_df.shape

(8760, 14)

There are total 8760 data points , 13 features and one target output i.e. Rented Bike Count

In [7]:
#summary
bike_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Date                       8760 non-null   object 
 1   Rented Bike Count          8760 non-null   int64  
 2   Hour                       8760 non-null   int64  
 3   Temperature(°C)            8760 non-null   float64
 4   Humidity(%)                8760 non-null   int64  
 5   Wind speed (m/s)           8760 non-null   float64
 6   Visibility (10m)           8760 non-null   int64  
 7   Dew point temperature(°C)  8760 non-null   float64
 8   Solar Radiation (MJ/m2)    8760 non-null   float64
 9   Rainfall(mm)               8760 non-null   float64
 10  Snowfall (cm)              8760 non-null   float64
 11  Seasons                    8760 non-null   object 
 12  Holiday                    8760 non-null   object 
 13  Functioning Day            8760 non-null   objec

Numerical as well as categorical variables are present. There are no null values in any column.Date has datatype named object so we need to change it to date datatype.


In [8]:
#creating function to change data type of date
def get_date(Date):
  date_obj = dt.datetime.strptime(Date , "%d/%m/%Y")
  date_obj = pd.to_datetime(date_obj, format = "%Y-%m-%d")
  return date_obj


In [9]:
 #applying function
 bike_df["Date"] = bike_df["Date"].apply(get_date)

In [10]:
# checking for duplicate values
bike_df.duplicated().sum()

0

In [11]:
# checking for missing values
bike_df.isnull().sum()

Date                         0
Rented Bike Count            0
Hour                         0
Temperature(°C)              0
Humidity(%)                  0
Wind speed (m/s)             0
Visibility (10m)             0
Dew point temperature(°C)    0
Solar Radiation (MJ/m2)      0
Rainfall(mm)                 0
Snowfall (cm)                0
Seasons                      0
Holiday                      0
Functioning Day              0
dtype: int64

There are no duplicate as  well as missing values.

##Exploratory Data Analysis